File: app/src/main/java/io/mgba/Model/Library.java
package io.mgba.Model;

import com.annimon.stream.Stream;
import java.io.File;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import javax.inject.Inject;
import dagger.Lazy;
import io.mgba.Constants;
import io.mgba.Data.Database.Game;
import io.mgba.Data.Remote.DTOs.GameJSON;
import io.mgba.Data.Remote.Interfaces.IRequest;
import io.mgba.Model.IO.Decoder;
import io.mgba.Model.IO.FilesManager;
import io.mgba.Model.Interfaces.IDatabase;
import io.mgba.Model.Interfaces.IFilesManager;
import io.mgba.Model.Interfaces.ILibrary;
import io.mgba.Utils.IDependencyInjector;
import io.mgba.Utils.IDeviceManager;
import io.mgba.mgba;
import io.reactivex.Single;
import io.reactivex.annotations.NonNull;

public class Library implements ILibrary {
    @Inject IDatabase database;
    @Inject IFilesManager filesService;
    @Inject IDeviceManager deviceManager;
    @Inject Lazy<IRequest> webManager;

    public Library(@NonNull IDependencyInjector dependencyInjector) {
        dependencyInjector.inject(this);
    }

    public Library(@NonNull IDatabase database, @NonNull IFilesManager filesService,
                   @NonNull IDeviceManager deviceManager, IRequest webManager) {
        this.database = database;
        this.filesService = filesService;
        this.deviceManager = deviceManager;
        this.webManager = () -> webManager;
    }

    @Override
    public Single<List<Game>> prepareGames(int platform) {
        Single<List<Game>> ret =  Single.create(subscriber -> {

            if(platform == Constants.PLATFORM_GBC
                    || platform == Constants.PLATFORM_GBA
                    || platform == Constants.PLATFORM_FAVS){

                List<Game> games = database.getGamesForPlatform(platform);

                subscriber.onSuccess(games);

            }else{
                subscriber.onSuccess(new LinkedList<>());
            }
        });

        return ret.doOnError(mgba::report);
    }

    @Override
    public Single<List<Game>> query(String query) {
        Single<List<Game>> ret = Single.create(subscriber -> {

            if(query == null || query.length() == 0){
                subscriber.onSuccess(new LinkedList<>());
                return;
            }

            List<Game> games = database.queryForGames(query);

            subscriber.onSuccess(games);
        });

        return ret.doOnError(mgba::report);
    }

    @Override
    public Single<List<Game>> reloadGames(int... platform) {
        Single<List<Game>> ret = Single.create(subscriber -> {
            List<Game> games = database.getGames();
            removeGamesFromDatabase(games);

            final List<Game> updatedList = processNewGames(games);

            games.addAll(updatedList);

            Collections.sort(games, (o1, o2) -> o1.getName().compareTo(o2.getName()));
            subscriber.onSuccess(filter(platform, games));
        });

        return ret.doOnError(mgba::report);

    }

    @Override
    public Single<List<Game>> reloadGames(String path, int... platform) {
        filesService.setCurrentDirectory(path);
        return reloadGames(platform);
    }

    private void removeGamesFromDatabase(List<Game> games){
        Stream.of(games)
                .filter(g -> !g.getFile().exists())
                .forEach(g -> {
                    games.remove(g);
                    database.delete(g);}
                );
    }

    private List<Game> processNewGames(List<Game> games){
        return Stream.of(filesService.getGameList())
                .map(file -> new Game(file.getAbsolutePath(), getPlatform(file)))
                .filter(file -> games.size() == 0 || Stream.of(games).anyMatch(game -> !game.getFile().equals(file.getFile())))
                .map(game -> {
                    Stream.of(games).filter(otherGame -> otherGame.equals(game)).forEach(games::remove);

                    if (calculateMD5(game)) {
                        if(deviceManager.isConnectedToWeb())
                            searchWeb(game);
                        storeInDatabase(game);
                    }

                    return game;
                }).toList();
    }

    private List<Game> filter(int[] platform, List<Game> games){
        return Stream.of(games)
                     .filter(g -> {
                         for (int i = 0; i < platform.length; i++) {
                             if(platform[i] == g.getPlatform())
                                 return true;
                         }

                        return false;
                     })
                     .toList();
    }

    private int getPlatform(File file) {
        final String fileExtension = FilesManager.getFileExtension(file);

        if(Constants.PLATFORM_GBA_EXT.contains(fileExtension))
            return Constants.PLATFORM_GBA;

        return Constants.PLATFORM_GBC;
    }

    private void storeInDatabase(Game game){
        database.insert(game);
    }

    private void searchWeb(Game game){
        try {
            final GameJSON json = webManager.get()
                                      .getGameInformation(game.getMD5(), deviceManager.getDeviceLanguage())
                                      .execute()
                                      .body();

            if(json != null)
                copyInformation(game, json);
        } catch (Exception e) {
            mgba.report(e);
        }
    }

    private boolean calculateMD5(Game game) {
        String md5 = Decoder.getFileMD5ToString(game.getFile());
        game.setMD5(md5);

        return md5 != null;
    }

    private void copyInformation(Game game, GameJSON json){
        game.setName(json.getName());
        game.setDescription(json.getDescription());
        game.setDeveloper(json.getDeveloper());
        game.setGenre(json.getGenre());
        game.setReleased(json.getReleased());
        game.setCoverURL(json.getCover());
    }
}

File: app/src/main/java/io/mgba/model/Library.kt
package io.mgba.model

import com.annimon.stream.Stream
import java.io.File
import java.util.LinkedList
import javax.inject.Inject
import dagger.Lazy
import io.mgba.Constants
import io.mgba.data.database.Game
import io.mgba.data.remote.dtos.GameJSON
import io.mgba.data.remote.interfaces.IRequest
import io.mgba.model.io.Decoder
import io.mgba.model.io.FilesManager
import io.mgba.model.interfaces.IDatabase
import io.mgba.model.interfaces.IFilesManager
import io.mgba.model.interfaces.ILibrary
import io.mgba.utilities.IDependencyInjector
import io.mgba.utilities.IDeviceManager
import io.mgba.mgba
import io.reactivex.Single
import io.reactivex.annotations.NonNull

class Library : ILibrary {
    @Inject internal lateinit var database: IDatabase
    @Inject internal lateinit var filesService: IFilesManager
    @Inject internal lateinit var deviceManager: IDeviceManager
    @Inject internal lateinit var webManager: Lazy<IRequest>

    constructor(@NonNull dependencyInjector: IDependencyInjector) {
        dependencyInjector.inject(this)
    }

    constructor(@NonNull database: IDatabase, @NonNull filesService: IFilesManager,
                @NonNull deviceManager: IDeviceManager, webManager: IRequest) {
        this.database = database
        this.filesService = filesService
        this.deviceManager = deviceManager
        this.webManager = Lazy{ webManager }
    }

    override fun prepareGames(platform: Int): Single<List<Game>> {
        val ret = Single.create<List<Game>> { subscriber ->

            if (platform == Constants.PLATFORM_GBC
                    || platform == Constants.PLATFORM_GBA
                    || platform == Constants.PLATFORM_FAVS) {

                val games = database.getGamesForPlatform(platform)

                subscriber.onSuccess(games)

            } else {
                subscriber.onSuccess(LinkedList())
            }
        }

        return ret.doOnError { mgba.report(it) }
    }

    override fun query(query: String): Single<List<Game>> {
        val ret = Single.create<List<Game>> Single@{ subscriber ->

            if (query.isEmpty()) {
                subscriber.onSuccess(LinkedList())
                return@Single
            }

            val games = database.queryForGames(query)

            subscriber.onSuccess(games)
        }

        return ret.doOnError { mgba.report(it) }
    }

    override fun reloadGames(vararg platform: Int): Single<List<Game>> {
        val ret = Single.create<List<Game>> { subscriber ->
            val games = database.games.toMutableList()
            removeGamesFromDatabase(games)

            val updatedList = processNewGames(games)

            games.addAll(updatedList)
            games.sortWith(Comparator { o1, o2 -> o1.getName()!!.compareTo(o2.getName()!!) })
            subscriber.onSuccess(filter(platform, games))
        }

        return ret.doOnError { mgba.report(it) }

    }

    override fun reloadGames(path: String, vararg platform: Int): Single<List<Game>> {
        filesService.currentDirectory = path
        return reloadGames(*platform)
    }

    private fun removeGamesFromDatabase(games: MutableList<Game>) {
        Stream.of(games)
                .filter { g -> !g.file!!.exists() }
                .forEach { g ->
                    games.remove(g)
                    database.delete(g)
                }
    }

    private fun processNewGames(games: MutableList<Game>): List<Game> {
        return Stream.of(filesService.gameList)
                .map { file -> Game(file.absolutePath, getPlatform(file)) }
                .filter { file -> games.isEmpty() || Stream.of(games).anyMatch { game -> game.file != file.file } }
                .map { game ->
                    Stream.of(games).filter { otherGame -> otherGame == game }
                                    .forEach { games.remove(it) }

                    if (calculateMD5(game)) {
                        if (deviceManager.isConnectedToWeb)
                            searchWeb(game)
                        storeInDatabase(game)
                    }

                    game
                }.toList()
    }

    private fun filter(platform: IntArray, games: List<Game>): List<Game> {
        return Stream.of(games)
                .filter Stream@{ g ->
                    for (i in platform.indices) {
                        if (platform[i] == g.platform)
                            return@Stream true
                    }
                    false
                }
                .toList()
    }

    private fun getPlatform(file: File): Int {
        val fileExtension = FilesManager.getFileExtension(file)

        return if (Constants.PLATFORM_GBA_EXT.contains(fileExtension)) Constants.PLATFORM_GBA else Constants.PLATFORM_GBC

    }

    private fun storeInDatabase(game: Game) {
        database.insert(game)
    }

    private fun searchWeb(game: Game) {
        try {
            val json = webManager.get()
                    .getGameInformation(game.mD5!!, deviceManager.deviceLanguage)
                    .execute()
                    .body()

            if (json != null)
                copyInformation(game, json)
        } catch (e: Exception) {
            mgba.report(e)
        }

    }

    private fun calculateMD5(game: Game): Boolean {
        val md5 = Decoder.getFileMD5ToString(game.file!!)
        game.mD5 = md5

        return md5 != null
    }

    private fun copyInformation(game: Game, json: GameJSON) {
        game.setName(json.name)
        game.description = json.description
        game.developer = json.developer
        game.genre = json.genre
        game.released = json.released
        game.coverURL = json.cover
    }
}
